src/VisualStudio/Core/Def/ProjectSystem/SdkAnalyzerAssemblyRedirector.cs

```diff
@@ -0,0 +1,190 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.IO;
+using System.Text.Json;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Workspaces.AnalyzerRedirecting;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+
+// Example:
+// FullPath: "C:\Program Files\dotnet\packs\Microsoft.WindowsDesktop.App.Ref\8.0.8\analyzers\dotnet\System.Windows.Forms.Analyzers.dll"
+// ProductVersion: "8.0.8"
+// PathSuffix: "analyzers\dotnet"
+using AnalyzerInfo = (string FullPath, string ProductVersion, string PathSuffix);
+
+namespace Microsoft.VisualStudio.LanguageServices.ProjectSystem;
+
+/// <summary>
+/// See <see href="https://github.com/dotnet/sdk/blob/main/documentation/general/analyzer-redirecting.md"/>.
+/// </summary>
+[Export(typeof(IAnalyzerAssemblyRedirector))]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal sealed class SdkAnalyzerAssemblyRedirector(SVsServiceProvider serviceProvider) : SdkAnalyzerAssemblyRedirectorCore(
+    Path.GetFullPath(Path.Combine(AppDomain.CurrentDomain.BaseDirectory, @"CommonExtensions\Microsoft\DotNet")),
+    (IVsActivityLog)serviceProvider.GetService(typeof(SVsActivityLog)));
+
+/// <summary>
+/// Core functionality of <see cref="SdkAnalyzerAssemblyRedirector"/> extracted for testing.
+/// </summary>
+internal class SdkAnalyzerAssemblyRedirectorCore : IAnalyzerAssemblyRedirector
+{
+    private readonly IVsActivityLog? _log;
+
+    private readonly bool _enabled;
+
+    private readonly string? _insertedAnalyzersDirectory;
+
+    /// <summary>
+    /// Map from analyzer assembly name (file name without extension) to a list of matching analyzers.
+    /// </summary>
+    private readonly ImmutableDictionary<string, List<AnalyzerInfo>> _analyzerMap;
+
+    public SdkAnalyzerAssemblyRedirectorCore(string? insertedAnalyzersDirectory, IVsActivityLog? log = null)
+    {
+        _log = log;
+        var enable = Environment.GetEnvironmentVariable("DOTNET_ANALYZER_REDIRECTING");
+        _enabled = !"0".Equals(enable, StringComparison.OrdinalIgnoreCase) && !"false".Equals(enable, StringComparison.OrdinalIgnoreCase);
+        _insertedAnalyzersDirectory = insertedAnalyzersDirectory;
+        _analyzerMap = CreateAnalyzerMap();
```

Comment:
Is there a reason we can't make this lazy for the first use? We're OK with System.Lazy here so the change can be trivial.